### PR TITLE
[Feature](bangc-ops): add new op roiawarepool3d.

### PR DESCRIPTION
**Pull request by ZhangLearning**
Thanks for your contribution and we appreciate it a lot. 
 
## 1. Motivation

add new op roiaware_pool3d_forward and roiaware_pool3d_backward

## 2. Modification

1) mlu-ops/bangc-ops/kernels/roiaware_pool3d/*
2) mlu-ops/bangc-ops/mlu_op.h
3) mlu-ops/bangc-ops/mlu_op_kernel.h
5) mlu-ops/bangc-ops/test/mlu_op_gtest/pb_gtest/mlu_op_test_proto/*
6) mlu-ops/bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roiaware_pool3d_forward/*
6) mlu-ops/bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roiaware_pool3d_backward/*
8) mlu-ops/docs/bangc-docs/user_guide/9_operators/*
9) mlu-ops/docs/bangc-docs/design_docs/roiaware_pool3d_forward/*
9) mlu-ops/docs/bangc-docs/design_docs/roiaware_pool3d_backward/*

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For dynamic threshold standard details.

#### 3.1.2 Operator Scheme checklist

|     No.      |           Details            |      Check Results       |
|----------------|---------------------------|---------------------|
|        1       |          Supported hardware         | MLU270 <br> MLU290 <br>MLU370<br>MLU590 |
|        2       |          Job types          |     U1     |
|        3       |         Layouts            | ARRAY   |
|        4       |         Whether multi-dimensions are supported              |    X     |
|        5       |          Whether element zero is supported             |    X      |
|        6       |         Data type(half/float)       |         half / float         |
|        7      |        Whether there is size limit           |       |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [x] Data type test
- [x] Multidimensional tensor test
- [x] Layout test
- [x] Different size/integer remainder end segment/alignment misalignment test
- [x] Zero dimensional tensor test/zero element test
- [x] stability test
- [x] Multiple platform test
- [x] Gen_case module test
- [x] Nan/INF tests 
- [ ] Bug fix tests
- [x] For memory leak check details, see[GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).
- [x] For code coverage check details, see: [GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).
- [x] For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Standard](../docs/MLU-OPS-Performance-Acceptance-Standard.md).

#### 3.1.4 Parameter Check

When a new operator is submitted, the test points are given and the test results are stated.

| Test Point         | Acceptance Standard | Test Result (Error Message) |
| -------------- | -------- | -------------------- |
| Whether it conforms to the operator restriction | Normal error |                      |
| Whether illegal parameters are passed  | Normal error |                      |

### 3.3 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

平台：MLU370
[----------] Global test environment tear-down
[ SUMMARY  ] Total 32 cases of 1 op(s).
ALL PASSED.
[==========] 32 test cases from 1 test suite ran. (5229 ms total)
[  PASSED  ] 32 test cases.

平台：MLU590
[----------] Global test environment tear-down
[ SUMMARY  ] Total 32 cases of 1 op(s).
ALL PASSED.
[==========] 32 test cases from 1 test suite ran. (5145 ms total)
[  PASSED  ] 32 test cases.
/tmp/workspace/mluops_develop/mluops_shell

### CODE REVIEW

File: bangc-ops/mlu_op.h [hunk 1]
```diff
@@ -3776,13 +3776,266 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
     mluOpHandle_t handle, const mluOpPointerMode_t pointer_mode,
     const void *value, const mluOpTensorDescriptor_t output_desc, void *output);
 
+
+// Group:RoiawarePool3d
+/*!
+ * @brief Returns in \b workspace_size the size of the MLU memory that is used as an extra
+ * workspace to optimize the RoiawarePool3dForward operation.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3783]: ::RoiawarePool3dForward   前面加英文冒号，增加链接

File: bangc-ops/mlu_op.h [hunk 2]
```diff
@@ -3776,13 +3776,266 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ *
+ * The size of extra workspace is based on the given information of the RoiawarePool3dForward
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3785]: ::RoiawarePool3dForward

File: bangc-ops/mlu_op.h [hunk 3]
```diff
@@ -3776,13 +3776,266 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ * operation, including the input tensor descriptors \b pts_desc.
+ *
+ * @param[in] handle
+ * Handle to an MLUOP context that is used to manage MLU devices and queues in the
+ * RoiawarePool3dForward operation. For detailed information, see ::mluOpHandle_t.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3790]: ::RoiawarePool3dForward

File: bangc-ops/mlu_op.h [hunk 4]
```diff
@@ -3776,13 +3776,266 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ * @param[in] rois_desc
+ * Descriptor of rois, containing the dimension and layout of the rois tensor.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3792]: The descriptor of ROI, which contains the dimension and layout of the ROI tensor.

File: bangc-ops/mlu_op.h [hunk 5]
```diff
@@ -2113,7 +2113,7 @@ mluOpPriorBox(mluOpHandle_t handle,
  *  Descriptor of rois tensor, containing dimension and the layout of rois.
  *  For detailed information, see ::mluOpTensorDescriptor_t.
  *  @param[in] rois
- *  Pointer to the MLU memory that stores the rois tensor. \b rois[1] consists of
+ *  Pointer to the MLU memory that stores the ROI tensor. \b rois[1] consists of
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:2116]: rois 这里维持原来的吧，我理解错了

File: bangc-ops/mlu_op.h [hunk 6]
```diff
@@ -2218,7 +2218,7 @@ mluOpPsRoiPoolForward(mluOpHandle_t handle,
  *  Descriptor of the rois tensor, which contains the dimension and the layout
  *  of rois tensor. For detailed information, see ::mluOpTensorDescriptor_t.
  *  @param[in] rois
- *  Pointer to the MLU memory that stores the rois tensor.
+ *  Pointer to the MLU memory that stores the ROI tensor.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:2221]: 维持原来的rois

File: bangc-ops/mlu_op.h [hunk 7]
```diff
@@ -2310,7 +2310,7 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
  * The descriptor of rois tensor, which contains dimension and the layout of rois.
  * For detailed information, see ::mluOpTensorDescriptor_t.
  * @param[in] rois
- * Pointer to the MLU memory that stores rois tensors. \b rois[i] consists of [batch_id,
+ * Pointer to the MLU memory that stores ROI tensors. \b rois[i] consists of [batch_id,
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:2313]: 维持原来的

File: bangc-ops/mlu_op.h [hunk 8]
```diff
@@ -2418,7 +2418,7 @@ mluOpRoiAlignRotatedForward(mluOpHandle_t handle,
  * The descriptor of rois tensor, which contains dimension and the layout of rois.
  * For detailed information, see ::mluOpTensorDescriptor_t.
  * @param[in] rois
- * Pointer to the MLU memory that stores rois tensors. \b rois[i] consists
+ * Pointer to the MLU memory that stores ROI tensors. \b rois[i] consists
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:2421]: 维持之前的

File: bangc-ops/mlu_op.h [hunk 9]
```diff
@@ -3776,13 +3776,266 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ * The size of extra workspace is based on the given information of the RoiawarePool3dForward
+ * operation, including the input tensor descriptors \b pts_desc.
+ *
+ * @param[in] handle
+ * Handle to an MLUOP context that is used to manage MLU devices and queues in the
+ * RoiawarePool3dForward operation. For detailed information, see ::mluOpHandle_t.
+ * @param[in] rois_desc
+ * Descriptor of rois, containing the dimension and layout of the rois tensor.
+ * @param[in] pts_desc
+ * Descriptor of pts, containing the dimension and layout of the pts tensor.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3794]: The descriptor of pts, which contains the dimension and layout of the pts tensor.

File: bangc-ops/mlu_op.h [hunk 10]
```diff
@@ -3776,13 +3776,266 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ * @param[in] pts_feature_desc
+ * Descriptor of pts_feature, containing the dimension and layout of the pts_feature tensor.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3796]: The  descriptor of pts_feature, which contains the dimension and layout of the pts_feature tensor.

File: bangc-ops/mlu_op.h [hunk 11]
```diff
@@ -3776,13 +3776,266 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ * @param[out] workspace_size
+ * Pointer to the returned size of the extra workspace in bytes that is used in the
+ * RoiawarePool3dForward operation.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3799]: ::RoiawarePool3dForward

File: bangc-ops/mlu_op.h [hunk 12]
```diff
@@ -3776,13 +3776,266 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+                                           const mluOpTensorDescriptor_t pts_feature_desc,
+                                           size_t *workspace_size);
+
+// Group:RoiawarePool3d
+/*!
+ * @param[in] handle
+ * Handle to an MLUOP context that is used to manage MLU devices and queues in
+ * ::mluOpRoiawarePool3dForward operation. For detailed information, see ::mluOpHandle_t.
+ * @param[in] pool_method
+ * Pooling method of Roiaware, 0 is 'maxpool', 1 is 'avgpool'. Default: 0.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3829]: The default value is 0.

File: bangc-ops/mlu_op.h [hunk 13]
```diff
@@ -3776,13 +3776,266 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ * @param[in] boxes_num
+ * An integer value which is the number of the rois.
+ * @param[in] pts_num
+ * An integer value which is the number of the pts.
+ * @param[in] channels
+ * An integer value which is the number of the pts feature of channels.
+ * @param[in] rois_desc
+ * Descriptor of rois tensor, containing the dimension and layout of the rois tensor.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3837]: The descriptor of rois tensor, which contains the dimension and layout of the rois tensor.

File: bangc-ops/mlu_op.h [hunk 14]
```diff
@@ -3776,13 +3776,266 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ * @param[in] rois
+ * Pointer to the MLU memory that stores the ROI tensor.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3839]: ROI > rois

File: bangc-ops/mlu_op.h [hunk 15]
```diff
@@ -3776,13 +3776,266 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ * @param[in] pts_desc
+ * Descriptor of pts tensor, containing the dimension and layout of the pts tensor.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3841]: The descriptor of pts tensor, which contains the dimension and layout of the pts tensor.

File: bangc-ops/mlu_op.h [hunk 16]
```diff
@@ -3776,13 +3776,266 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ * @param[in] pts
+ * Pointer to the MLU memory that stores the pts tensor.
+ * @param[in] pts_feature_desc
+ * Descriptor of pts_feature tensor, containing the dimension and layout of the pts_feature tensor.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3845]: The descriptor of pts_feature tensor, which contains the dimension and layout of the pts_feature tensor.

File: bangc-ops/mlu_op.h [hunk 17]
```diff
@@ -3776,13 +3776,266 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ * @param[in] pts_feature
+ * Pointer to the MLU memory that stores the pts_feature tensor.
+ * @param[in] workspace
+ * Pointer to the MLU memory that is used as an extra workspace for the
+ * RoiawarePool3dForward operation.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3850]: ::RoiawarePool3dForward

File: bangc-ops/mlu_op.h [hunk 18]
```diff
@@ -3776,13 +3776,266 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ * @param[in] workspace_size
+ * The size of the extra workspace in bytes that needs to be used in
+ * the RoiawarePool3dForward operation. You can get the size of the workspace with
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3853]: ::RoiawarePool3dForward

File: bangc-ops/mlu_op.h [hunk 19]
```diff
@@ -3776,13 +3776,266 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ * the ::mluOpGetCosineSimilarityWorkspaceSize function.
+ * @param[in] max_pts_each_voxel
+ * The maximum number of points each voxel. An integer value which is the dimension of the pts_idx_of_voxels.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3856]: per each voxel

File: bangc-ops/mlu_op.h [hunk 20]
```diff
@@ -3776,13 +3776,266 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ * @param[in] out_x
+ * An integer value which is the dimension of the pooled_features.
+ * @param[in] out_y
+ * An integer value which is the dimension of the pooled_features.
+ * @param[in] out_z
+ * An integer value which is the dimension of the pooled_features.
+ * @param[in] argmax_desc
+ * Descriptor of argmax tensor, containing the dimension and layout of argmax tensor.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3864]: The descriptor of argmax tensor, which contains the dimension and layout of argmax tensor.

File: bangc-ops/mlu_op.h [hunk 21]
```diff
@@ -3776,13 +3776,266 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ * @param[out] argmax
+ * Pointer to the MLU memory that stores the argmax tensor.
+ * @param[in] pts_idx_of_voxels_desc
+ * Descriptor of pts_idx_of_voxels tensor, containing the dimension of pts_idx_of_voxels tensor.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3868]: The descriptor of pts_idx_of_voxels tensor, which contains the dimension of pts_idx_of_voxels tensor.

File: bangc-ops/mlu_op.h [hunk 22]
```diff
@@ -3776,13 +3776,266 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ * @param[out] pts_idx_of_voxels
+ * Pointer to the MLU memory that stores the pts_idx_of_voxels tensor.
+ * @param[in] pooled_features_desc
+ * Descriptor of pooled_features tensor, containing the dimension of pooled_features tensor.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3872]: The descriptor of pooled_features tensor, which contains the dimension of pooled_features tensor.

File: bangc-ops/mlu_op.h [hunk 23]
```diff
@@ -3776,13 +3776,266 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ * @param[out] pooled_features
+ * Pointer to the MLU memory that stores the pooled_features tensor.
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3876]: 去掉句尾的句号

File: bangc-ops/mlu_op.h [hunk 24]
```diff
@@ -3776,13 +3776,266 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ * Descriptor of argmax tensor, containing the dimension and layout of argmax tensor.
+ * @param[out] argmax
+ * Pointer to the MLU memory that stores the argmax tensor.
+ * @param[in] grad_out_desc
+ * Descriptor of grad_out tensor, containing the dimension of grad_out tensor.
+ * @param[out] grad_out
+ * Pointer to the MLU memory that stores the grad_out tensor.
+ * @param[in] workspace
+ * Pointer to the MLU memory that is used as an extra workspace for the
+ * RoiawarePool3dBackward operation.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3971]: ::RoiawarePool3dBackward

File: bangc-ops/mlu_op.h [hunk 25]
```diff
@@ -3776,13 +3776,266 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ * @param[in] workspace_size
+ * The size of the extra workspace in bytes that needs to be used in
+ * the RoiawarePool3dBackward operation. You can get the size of the workspace with
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3974]: ::RoiawarePool3dBackward

File: bangc-ops/mlu_op.h [hunk 26]
```diff
@@ -3776,13 +3776,266 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ * @par Scale Limitation
+ * - The shape of \b pts_idx_of_voxels should be [boxes_num, out_x, out_y, out_z, max_pts_each_voxel].
+ * - The shape of \b argmax should be [boxes_num, out_x, out_y, out_z, channels].
+ * - The shape of \b grad_out should be [boxes_num, out_x, out_y, out_z, channels].
+ * - The shape of \b grad_in should be [pts_num, channels].
+ *
+ * @par API Dependency
+ * - None.
+ *
+ * @par Performance Optimization
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4000]: 这个字段之前头文件没有，删除

File: bangc-ops/mlu_op.h [hunk 27]
```diff
@@ -3776,13 +3776,266 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ * - The shape of \b pts should be [pts_num, 3].
+ * - The shape of \b pts_feature should be [pts_num, channels].
+ * - The shape of \b argmax should be [boxes_num, out_x, out_y, out_z, channels].
+ * - The shape of \b pts_idx_of_voxels should be [boxes_num, out_x, out_y, out_z, max_pts_each_voxel].
+ * - The shape of \b pooled_features should be [boxes_num, out_x, out_y, out_z, channels].
+ *
+ * @par API Dependency
+ * - None.
+ *
+ * @par Performance Optimization
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3898]: 这个字段之前头文件没有，删除

**Review by GraceGuanH — APPROVED** [review]

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roiaware_pool3d_forward/roiaware_pool3d_forward.h [hunk 28]
```diff
@@ -0,0 +1,58 @@
...
+
+class RoiawarePool3dForwardExecutor : public Executor {
+ public:
+  RoiawarePool3dForwardExecutor() {}
+  ~RoiawarePool3dForwardExecutor() { workspaceFree(); }
+  void paramCheck() override;
+  void workspaceMalloc() override;
+  void workspaceFree() override;
+  void compute() override;
+  void checkDescParam();
```
Comment (tudejiang79) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roiaware_pool3d_forward/roiaware_pool3d_forward.h:27]: 放在private中

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roiaware_pool3d_forward/roiaware_pool3d_forward.h [hunk 29]
```diff
@@ -0,0 +1,58 @@
...
+  void *dev_argmax_ = NULL;
+  void *dev_pts_idx_of_voxels_ = NULL;
+  void *dev_pooled_features_ = NULL;
+  mluOpTensorDescriptor_t desc_rois_ = NULL;
+  mluOpTensorDescriptor_t desc_pts_ = NULL;
+  mluOpTensorDescriptor_t desc_pts_feature_ = NULL;
+  mluOpTensorDescriptor_t desc_argmax_ = NULL;
+  mluOpTensorDescriptor_t desc_pts_idx_of_voxels_ = NULL;
+  mluOpTensorDescriptor_t desc_pooled_features_ = NULL;
+  size_t workspace_size_ = 0;
```
Comment (tudejiang79) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roiaware_pool3d_forward/roiaware_pool3d_forward.h:54]: 类属性要初始化一起都初始化

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roiaware_pool3d_forward/roiaware_pool3d_forward.cpp [hunk 30]
```diff
@@ -0,0 +1,400 @@
...
+ *************************************************************************/
+#include "roiaware_pool3d_forward.h"
+
+#include <algorithm>
+#include <string>
+
+namespace mluoptest {
+
+static void lidarToLocalCoords(float shift_x, float shift_y, float rz,
+                               float &local_x, float &local_y) {
```
Comment (tudejiang79) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roiaware_pool3d_forward/roiaware_pool3d_forward.cpp:20]: 加const

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roiaware_pool3d_forward/roiaware_pool3d_forward.cpp [hunk 31]
```diff
@@ -0,0 +1,400 @@
...
+
+void RoiawarePool3dForwardExecutor::paramCheck() {
+  VLOG(4) << "RoiawarePool3dForwardExecutor::paramCheck() Begin.";
+  GTEST_CHECK(parser_->getInputNum() == 3);
+  GTEST_CHECK(parser_->getOutputNum() == 3);
+  if (!parser_->getProtoNode()->has_roiaware_pool3d_forward_param()) {
+    LOG(ERROR) << "[GTEST_ROIAWARE_POOL3D_FORWARD] Missing "
+                  "roiaware_pool3d_forward param.";
+    throw std::invalid_argument(std::string(__FILE__) + " +" +
+                                std::to_string(__LINE__));
```
Comment (tudejiang79) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roiaware_pool3d_forward/roiaware_pool3d_forward.cpp:255]: 这地方改为用GTEST_CHECK()

File: bangc-ops/mlu_op_kernel.h [hunk 32]
```diff
@@ -146,6 +146,48 @@ void MLUOP_WIN_API mluOpBlockKernelPsRoiPoolBackwardFloat(
     const int output_dim, const int rois_num, const int rois_offset,
     const float spatial_scale);
 
+void MLUOP_WIN_API mluOpUnionKernelPtsIdxOfVoxelsHalf(
```
Comment (tudejiang79) [inline bangc-ops/mlu_op_kernel.h:149]: 少注释

File: bangc-ops/kernels/roiaware_pool3d/roiaware_pool3d.cpp [hunk 33]
```diff
@@ -0,0 +1,614 @@
...
+  mluOpStatus_t status = MLUOP_STATUS_BAD_PARAM;
+  status = kernelPtsIdxOfVoxelsPolicyFunc(handle, pts_idx_of_voxels_desc,
+                                          &k_dim, &k_type);
+  if (MLUOP_STATUS_SUCCESS != status) {
+    GEN_CASE_END();
+    return status;
+  }
+
+  int core_dim = mluop::runtime::getClusterLimitCapability(handle);
+  VLOG(5) << "[mluOpRoiawarePool3dBackward] Launch Kernel "
```
Comment (wushaoqiang2168) [inline bangc-ops/kernels/roiaware_pool3d/roiaware_pool3d.cpp:391]: mluOpRoiawarePool3dBackward ？下同

File: bangc-ops/kernels/roiaware_pool3d/roiaware_pool3d_union1.mlu [hunk 34]
```diff
@@ -0,0 +1,732 @@
...
+    const int max_pts_each_voxel, const int out_x, const int out_y,
+    const int out_z, const void *rois, const void *pts,
+    void *pts_idx_of_voxels) {
+  MLUMultiKernelPtsIdxOfVoxels<<<k_dim, k_type, queue>>>(
+      pool_method, boxes_num, pts_num, max_pts_each_voxel, out_x, out_y, out_z,
+      (half *)rois, (half *)pts, (int *)pts_idx_of_voxels);
+}
+
+void MLUOP_WIN_API mluOpUnionKernelPtsIdxOfVoxelsFloat(
+
```
Comment (wushaoqiang2168) [inline bangc-ops/kernels/roiaware_pool3d/roiaware_pool3d_union1.mlu:525]: 多一行空行

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roiaware_pool3d_backward/roiaware_pool3d_backward.cpp [hunk 35]
```diff
@@ -0,0 +1,219 @@
+/*************************************************************************
+ * Copyright (C) [2019-2022] by Cambricon, Inc.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
```
Comment (wushaoqiang2168) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roiaware_pool3d_backward/roiaware_pool3d_backward.cpp:10]: license 修改下

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roiaware_pool3d_backward/roiaware_pool3d_backward.h [hunk 36]
```diff
@@ -0,0 +1,50 @@
...
+ * Copyright (C) [2019-2022] by Cambricon, Inc.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *************************************************************************/
```
Comment (wushaoqiang2168) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roiaware_pool3d_backward/roiaware_pool3d_backward.h:11]: license更新下

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roiaware_pool3d_forward/roiaware_pool3d_forward.cpp [hunk 37]
```diff
@@ -0,0 +1,400 @@
...
+ * Copyright (C) [2019-2022] by Cambricon, Inc.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *************************************************************************/
```
Comment (wushaoqiang2168) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roiaware_pool3d_forward/roiaware_pool3d_forward.cpp:11]: license

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roiaware_pool3d_forward/roiaware_pool3d_forward.h [hunk 38]
```diff
@@ -0,0 +1,58 @@
...
+  void workspaceFree() override;
+  void compute() override;
+  void checkDescParam();
+  void cpuCompute() override;
+  int64_t getTheoryOps() override;
+
+ private:
+  void initData();
+  void printDataInfo();
+  int pool_method_;
```
Comment (wushaoqiang2168) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roiaware_pool3d_forward/roiaware_pool3d_forward.h:34]: 类成员变量初始化下吧

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roiaware_pool3d_forward/roiaware_pool3d_forward.cpp [hunk 39]
```diff
@@ -0,0 +1,400 @@
...
+              desc_pooled_features_->dtype == MLUOP_DTYPE_HALF);
+  GTEST_CHECK(desc_argmax_->dtype == MLUOP_DTYPE_INT32);
+
+  VLOG(4) << "RoiawarePool3dForwardExecutor::checkDescParam() End.";
+}
+
+void RoiawarePool3dForwardExecutor::cpuCompute() {
+  VLOG(4) << "RoiawarePool3dForwardExecutor::cpuCompute() Begin.";
+
+  checkDescParam();
```
Comment (wushaoqiang2168) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roiaware_pool3d_forward/roiaware_pool3d_forward.cpp:354]: 没必要在这里在做一次防呆check吧

File: bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roiaware_pool3d_backward/roiaware_pool3d_backward.h [hunk 40]
```diff
@@ -0,0 +1,50 @@
...
+  ~RoiawarePool3dBackwardExecutor() {}
+  void paramCheck() override;
+  void compute() override;
+  void cpuCompute() override;
+  int64_t getTheoryOps() override;
+
+ private:
+  void initData();
+  void printDataInfo();
+  int pool_method_;
```
Comment (wushaoqiang2168) [inline bangc-ops/test/mlu_op_gtest/pb_gtest/src/zoo/roiaware_pool3d_backward/roiaware_pool3d_backward.h:31]: 初始化下

**Review by wushaoqiang2168 — APPROVED** [review]

**Review by tudejiang79 — APPROVED** [review]

File: docs/bangc-docs/design_docs/roiaware_pool3d_backward/roiaware_pool3d_backward.md [hunk 41]
```diff
@@ -0,0 +1,349 @@
+# roiaware_pool3d_backward 算子开发设计方案
+
+- #### 文档基本信息
+
+| 项目名称    | Training Solution        |
```
Comment (WangYuan-cn) [inline docs/bangc-docs/design_docs/roiaware_pool3d_backward/roiaware_pool3d_backward.md:5]: 改为mlu-ops模板

**Review by WangYuan-cn — APPROVED** [review]

**Pull request merged**